test/integration/controller_test.go

```diff
@@ -256,6 +252,146 @@ func TestBasicFlows(t *testing.T) {
+// do it here or in a follow-up; reviewer's choice.
+func TestProvisionFailure(t *testing.T) {
+	_, catalogClient, _, _, _, shutdownServer := newTestController(t, fakeosb.FakeClientConfiguration{
+		CatalogReaction: &fakeosb.CatalogReaction{
+			Response: &osb.CatalogResponse{
+				Services: []osb.Service{
+					{
+						Name:        testServiceClassName,
+						ID:          "12345",
+						Description: "a test service",
+						Bindable:    true,
+						Plans: []osb.Plan{
+							{
+								Name:        testPlanName,
+								Free:        truePtr(),
+								ID:          "34567",
+								Description: "a test plan",
+							},
+						},
+					},
+				},
+			},
+		},
+		ProvisionReaction: &fakeosb.ProvisionReaction{
+			Error: osb.HTTPStatusCodeError{
+				StatusCode:   http.StatusConflict,
+				ErrorMessage: strPtr("OutOfQuota"),
+				Description:  strPtr("You're out of quota!"),
+			},
+		},
+		// no DeprovisionReaction is configured, so that the client will
+		// return an unexpected call error message if deprovision is called on
+		// the broker.
+	})
+	defer shutdownServer()
+
+	client := catalogClient.ServicecatalogV1alpha1()
+
+	broker := &v1alpha1.Broker{
+		ObjectMeta: metav1.ObjectMeta{Name: testBrokerName},
+		Spec: v1alpha1.BrokerSpec{
+			URL: testBrokerURL,
+		},
+	}
+
+	_, err := client.Brokers().Create(broker)
+	if nil != err {
+		t.Fatalf("error creating the broker %q (%q)", broker, err)
+	}
+
+	err = util.WaitForBrokerCondition(client,
+		testBrokerName,
+		v1alpha1.BrokerCondition{
+			Type:   v1alpha1.BrokerConditionReady,
+			Status: v1alpha1.ConditionTrue,
+		})
+	if err != nil {
+		t.Fatalf("error waiting for broker to become ready: %v", err)
+	}
+
+	err = util.WaitForServiceClassToExist(client, testServiceClassName)
+	if nil != err {
+		t.Fatalf("error waiting from ServiceClass to exist: %v", err)
+	}
+
+	// TODO: find some way to compose scenarios; extract method here for real
+	// logic for this test.
+
+	//-----------------
+
+	instance := &v1alpha1.Instance{
+		ObjectMeta: metav1.ObjectMeta{Namespace: testNamespace, Name: testInstanceName},
+		Spec: v1alpha1.InstanceSpec{
+			ServiceClassName: testServiceClassName,
+			PlanName:         testPlanName,
+			ExternalID:       testExternalID,
+		},
+	}
+
+	if _, err := client.Instances(testNamespace).Create(instance); err != nil {
+		t.Fatalf("error creating Instance: %v", err)
+	}
+
+	if err := util.WaitForInstanceCondition(client, testNamespace, testInstanceName, v1alpha1.InstanceCondition{
+		Type:   v1alpha1.InstanceConditionFailed,
+		Status: v1alpha1.ConditionTrue,
+	}); err != nil {
+		t.Fatalf("error waiting for instance to become failed: %v", err)
+	}
+
+	retInst, err := client.Instances(instance.Namespace).Get(instance.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("error getting instance %s/%s back", instance.Namespace, instance.Name)
+	}
+	if retInst.Spec.ExternalID != instance.Spec.ExternalID {
```

Comment:
I copied it from the existing test case and didn't want to start trimming things out, because i think there will be an easy way to extract methods and have fluent style tests.